include/nbl/builtin/glsl/blit/blit/blit.glsl

```diff
@@ -0,0 +1,177 @@
+	const vec3 minOutputPixelCenterOfWG = vec3(minOutputPixel)*scale + halfScale;
+	const ivec3 regionStartCoord = getMinKernelWindowCoord(minOutputPixelCenterOfWG, params.negativeSupport); // this can be negative, in which case we wrap
+
+	const uvec3 preloadRegion = params.preloadRegion;
+
+	for (uint virtualInvocation = gl_LocalInvocationIndex; virtualInvocation < preloadRegion.x * preloadRegion.y * preloadRegion.z; virtualInvocation += _NBL_GLSL_WORKGROUP_SIZE_)
+	{
+		const ivec3 inputPixelCoord = regionStartCoord + ivec3(linearIndexTo3DIndex(virtualInvocation, preloadRegion));
+
+		vec3 inputTexCoord = (inputPixelCoord + vec3(0.5f)) / params.inDim;
+		
+		const vec4 loadedData = nbl_glsl_blit_getData(inputTexCoord, gl_WorkGroupID.z);
+		for (uint ch = 0; ch < _NBL_GLSL_BLIT_OUT_CHANNEL_COUNT_; ++ch)
+			scratchShared[ch][virtualInvocation] = loadedData[ch];
+	}
+	barrier();
+
+	const uvec3 iterationRegions[3] = uvec3[]( uvec3(outputTexelsPerWG.x, preloadRegion.yz), uvec3(outputTexelsPerWG.yx, preloadRegion.z), outputTexelsPerWG.yxz );
+
+	uint read = 0;
+	for (uint axis = 0; axis < _NBL_GLSL_BLIT_DIM_COUNT_; ++axis)
+	{
+		const uvec3 iterationRegion = iterationRegions[axis];
+		for (uint virtualInvocation = gl_LocalInvocationIndex; virtualInvocation < iterationRegion.x * iterationRegion.y * iterationRegion.z; virtualInvocation += _NBL_GLSL_WORKGROUP_SIZE_)
+		{
+			const uvec3 virtualInvocationID = linearIndexTo3DIndex(virtualInvocation, iterationRegion);
+
+			uint outputPixel = virtualInvocationID.x;
+			if (axis == 2)
+				outputPixel = virtualInvocationID.z;
+			outputPixel += minOutputPixel[axis];
+
+			if (outputPixel >= params.outDim[axis])
+				break;
+
+			const int minKernelWindow = int(ceil((outputPixel + 0.5f) * scale[axis] - 0.5f + params.negativeSupport[axis]));
+
+			// Combined stride for the two non-blitting dimensions, tightly coupled and experimentally derived with/by `iterationRegion` above and the general order of iteration we use to avoid
+			// read bank conflicts.
+			uint combinedStride;
+			{
+				if (axis == 0)
+					combinedStride = virtualInvocationID.z * preloadRegion.y + virtualInvocationID.y;
+				else if (axis == 1)
+					combinedStride = virtualInvocationID.z * outputTexelsPerWG.x + virtualInvocationID.y;
+				else if (axis == 2)
+					combinedStride = virtualInvocationID.y * outputTexelsPerWG.y + virtualInvocationID.x;
+			}
+
+			uint offset = (minKernelWindow - regionStartCoord[axis]) + combinedStride*preloadRegion[axis];
+			const uint windowPhase = outputPixel % params.phaseCount[axis];
+
+			uint kernelWeightIndex;
+			if (axis == 0)
+				kernelWeightIndex = windowPhase * params.windowDim.x;
+			else if (axis == 1)
+				kernelWeightIndex = params.phaseCount.x * params.windowDim.x + windowPhase * params.windowDim.y;
+			else if (axis == 2)
+				kernelWeightIndex = params.phaseCount.x * params.windowDim.x + params.phaseCount.y * params.windowDim.y + windowPhase * params.windowDim.z;
```

Comment:
precompute the Y and Z axis offsets and put them in the push constants

Reply:
I think we have already exhausted the 128 byte _required_ push constant limit.

Reply:
things can be packed

Reply:
Resolved.